plugin/route53/route53.go

```diff
@@ -117,18 +117,28 @@ func (h *Route53) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg
 	m.SetReply(r)
 	m.Authoritative = true
 	var result file.Result
+
+	tmpMsg := new(dns.Msg)
+	var tmpResult file.Result
+	isNodata := false
+
 	for _, hostedZone := range z {
 		h.zMu.RLock()
 		m.Answer, m.Ns, m.Extra, result = hostedZone.z.Lookup(ctx, state, qname)
 		h.zMu.RUnlock()
 
-		// Take the answer if it's non-empty OR if there is another
-		// record type exists for this name (NODATA).
-		if len(m.Answer) != 0 || result == file.NoData {
+		if len(m.Answer) != 0 {
 			break
+		} else if result != file.NameError {
+			tmpMsg.Answer, tmpMsg.Ns, tmpMsg.Extra, tmpResult = m.Answer, m.Ns, m.Extra, result
+			isNodata = true
 		}
 	}
 
+	if isNodata && result != file.Success {
+		m.Answer, m.Ns, m.Extra, result = tmpMsg.Answer, tmpMsg.Ns, tmpMsg.Extra, tmpResult
+	}
```

Comment:
How about something like this?
I added some comments and changed some styles.
The concept itself isn't changed though.

```
		// If there is an answer, it will be returned immediately.
		if len(m.Answer) != 0 {
			break
		}

		// Only NXDOMAIN doesn't need to be stored.
		// Other cases such as NODATA, Delegation, SERVFAIL should be respected.
		if result != file.NameError {
			tmpMsg.Answer, tmpMsg.Ns, tmpMsg.Extra, tmpResult = m.Answer, m.Ns, m.Extra, result
			isNodata = true
		}
	}

	// If the first zone returns NODATA and the second zone does not return an answer,
	// the first NODATA message will be used.
	if isNodata && result != file.Success {
		m.Answer, m.Ns, m.Extra, result = tmpMsg.Answer, tmpMsg.Ns, tmpMsg.Extra, tmpResult
	}
```